File: src/server/file_server.ts
import path from "path";
import fs from "fs";
import mime from "mime-types";
import { Body, HTTPStatus, is_object, is_string, Message, Methods } from "./lib";
import Config from "./config";
import { Request, Response } from "express";
import { Express } from "express";

/**
 * Handle the requests of vuefinder
 */
export default class FileServer {
	private endpoint_map: Pick<
		Record<Methods, Record<string, (req: Request) => Message>>,
		"GET" | "POST"
	> = {
		// eslint-disable-next-line @typescript-eslint/naming-convention
		GET: {
			index: (req: Request) => this.get_files(req),
			preview: (req: Request) => this.get_preview(req),
			subfolders: (req: Request) => this.get_subfolders(req),
			download: (req: Request) => this.get_download(req)
		},
		// eslint-disable-next-line @typescript-eslint/naming-convention
		POST: {
			newfolder: (req: Request) => this.create_new_folder(req),
			rename: (req: Request) => this.rename_file(req),
			move: (req: Request) => this.rename_files(req),
			delete: (req: Request) => this.post_delete(req)
		}
	};

	/**
	 *
	 * @param app express-instance to use
	 */
	constructor(app: Express) {
		app.get("/api/storage/browse", (req: Request, res: Response) => {
			void this.get(req, res);
		});

		app.post("/api/storage/browse", (req: Request, res: Response) => {
			void this.post(req, res);
		});
	}

	/**
	 * handles all GET-requests
	 * @param req Request
	 * @param res Response
	 */
	get(req: Request, res: Response) {
		if (typeof req.query.q === "string") {
			if (this.endpoint_map.GET[req.query.q] !== undefined) {
				const message = this.endpoint_map.GET[req.query.q]?.(req);

				res.status(message.status);

				if (message.json !== undefined) {
					res.json(message.json);
				} else if (message.buffer !== undefined) {
					res.setHeader("Content-Type", "application/octet-stream");
					res.send(message.buffer);
				} else {
					res.send(message.message);
				}
			} else {
				res.status(HTTPStatus.NotImplemented).send();
			}
		} else {
			res.status(HTTPStatus.BadRequest).send();
		}
	}

	/**
	 * Handle all POST-requests
	 * @param req Request
	 * @param res Response
	 */
	post(req: Request, res: Response) {
		if (typeof req.query.q === "string") {
			if (this.endpoint_map.POST[req.query.q] !== undefined) {
				const message = this.endpoint_map.POST[req.query.q]?.(req);

				res.status(message.status);

				if (message.json !== undefined) {
					res.json(message.json);
				} else if (message.buffer !== undefined) {
					res.setHeader("Content-Type", "application/octet-stream");
					res.send(message.buffer);
				} else {
					res.send(message.message);
				}
			} else {
				res.status(HTTPStatus.NotImplemented).send();
			}
		} else {
			res.status(HTTPStatus.BadRequest).send();
		}
	}

	/**
	 * handle list-file-requests
	 * @param req Request
	 * @returns client-response-message
	 */
	protected get_files(req: Request): Message {
		const adapter =
			is_string(req.query.adapter) && req.query.adapter !== "null" ? req.query.adapter : "PUBLIC";

		const pth = this.extract_path(is_string(req.query.path) ? req.query.path : "/", adapter);
		const pth_local = Config.get_upload_dir(pth);

		const files = fs.readdirSync(pth_local);

		return {
			status: HTTPStatus.OK,
			json: {
				adapter,
				storages: [adapter],
				dirname: pth,
				files: files.map((ff_name) => {
					const ff = path.join(pth, ff_name);
					const ff_local = Config.get_upload_dir(ff);

					return this.to_vuefinder_resource(adapter, ff, fs.statSync(ff_local));
				})
			}
		};
	}

	/**
	 * send a preview
	 * @param req Request
	 * @returns client-response-message
	 */
	protected get_preview(req: Request): Message {
		if (typeof req.query.path === "string" && typeof req.query.adapter === "string") {
			return {
				status: HTTPStatus.OK,
				buffer: fs.readFileSync(
					Config.get_upload_dir(this.extract_path(req.query.path, req.query.adapter))
				)
			};
		} else {
			return { status: HTTPStatus.BadRequest };
		}
	}

	/**
	 * get the subdirectories of a path
	 * @param req Request
	 * @returns client-response-message
	 */
	protected get_subfolders(req: Request): Message {
		const query = req.query;
		if (is_string(query.adapter) && is_string(query.path)) {
			const adapter = query.adapter;
			const pth = this.extract_path(query.path, adapter);
			const pth_local = Config.get_upload_dir(pth);

			return {
				status: HTTPStatus.OK,
				json: {
					folders: fs
						.readdirSync(pth_local)
						.map((ff) => {
							const stats = fs.statSync(path.join(pth_local, ff));

							if (stats.isDirectory()) {
								return this.to_vuefinder_resource(adapter, path.join(pth, ff), stats);
							} else {
								return null;
							}
						})
						.filter((ff) => ff !== null)
				}
			};
		} else {
			return { status: HTTPStatus.BadRequest };
		}
	}

	/**
	 * downloads a file
	 * @param req Reqeust
	 * @returns client-response-message
	 */
	protected get_download(req: Request): Message {
		return this.get_preview(req);
	}

	/**
	 * create a new folder
	 * @param req Request
	 * @returns client-response-message
	 */
	protected create_new_folder(req: Request): Message {
		const body = req.body as Body;

		if (
			typeof req.query.adapter === "string" &&
			typeof req.query.path === "string" &&
			typeof body === "object" &&
			typeof body.name === "string"
		) {
			fs.mkdirSync(path.join(Config.server.upload_dir, req.query.path, body.name));

			return this.get_files(req);
		} else {
			return { status: HTTPStatus.NotImplemented };
		}
	}

	/**
	 * rename a file
	 * @param req Request
	 * @returns client-response-message
	 */
	protected rename_file(req: Request): Message {
		const query = req.query;
		const body = req.body as Body;

		if (is_string(query.adapter) && is_string(body.item) && is_string(body.name)) {
			const orig = Config.get_upload_dir(this.extract_path(body.item, query.adapter));
			const dest = path.join(path.dirname(orig), body.name);

			fs.renameSync(orig, dest);

			return this.get_files(req);
		} else {
			return { status: HTTPStatus.NotImplemented };
		}
	}

	/**
	 * moves files
	 * @param req Request
	 * @returns client-response-message
	 */
	protected rename_files(req: Request): Message {
		const body = req.body as Body;

		if (is_string(req.query.adapter) && is_string(body.item) && is_items_array(body.items)) {
			const adapter = req.query.adapter;
			const destination = Config.get_upload_dir(this.extract_path(body.item, adapter));

			body.items.forEach((ff) => {
				const orig = this.extract_path((ff as { path: string }).path, adapter);

				fs.renameSync(Config.get_upload_dir(orig), path.join(destination, path.basename(orig)));
			});

			return this.get_files(req);
		} else {
			return { status: HTTPStatus.BadRequest };
		}
	}

	/**
	 * delete files
	 * @param req Request
	 * @returns client-response-message
	 */
	protected post_delete(req: Request): Message {
		const query = req.query;
		const body = req.body as Body;

		if (is_string(query.adapter) && Array.isArray(body.items) && is_items_array(body.items)) {
			const adapter = query.adapter;

			body.items.forEach((ff) => {
				const local_path = Config.get_upload_dir(this.extract_path(ff.path, adapter));

				if (ff.type === "file") {
					fs.rmSync(local_path);
				} else {
					fs.rmdirSync(local_path, { recursive: true });
				}
			});

			return this.get_files(req);
		} else {
			return { status: HTTPStatus.BadRequest };
		}
	}

	/**
	 * remove the adapter from the path
	 * @param pth path with uri
	 * @param adapter adapter used in the uri
	 * @returns path without adapter
	 */
	private extract_path(pth: string, adapter: string): string {
		return pth.replace(`${adapter}:/`, "");
	}

	/**
	 * craft a vuefinder-resource for a file
	 * @param adapter adapter for the path
	 * @param pth path of the directory
	 * @param info file-info
	 * @returns vuefinder-resource
	 */
	private to_vuefinder_resource(adapter: string, pth: string, info: fs.Stats) {
		const mime_type = mime.lookup(pth);

		const pth_parse = path.parse(pth);

		if (pth[0] === "/") {
			pth = pth.slice(1);
		}

		const data = {
			type: info.isDirectory() ? "dir" : "file",
			path: `${adapter}:/${pth}`,
			visibility: "public",
			last_modified: info.mtime.getTime() / 1000,
			mime_type: mime_type !== false ? mime_type : "text/plain",
			extra_metadata: [],
			basename: pth_parse.base,
			extension: pth_parse.ext,
			storage: adapter,
			file_size: info.size
		};

		return data;
	}
}

/**
 * checks wether a unknown is an array of file-definitions ({ path: string; type: string; }[])
 * @param a unknown variable
 * @returns wether a is of type: { path: string; type: string; }[]
 */
function is_items_array(a: unknown): a is Array<{ path: string; type: string }> {
	return (
		Array.isArray(a) &&
		a.every((ele) => {
			const obj = ele as Record<string, unknown>;

			return is_object(ele) && !Array.isArray(ele) && is_string(obj.path) && is_string(obj.type);
		})
	);
}

File: fileServer.go
package main

import (
	"io/fs"
	"mime"
	"os"
	"path"
	"strings"

	"github.com/gofiber/fiber/v2"
)

func extractPath(p string, a string) string {
	return strings.Replace(p, a+"://", "", 1)
}

type VuefinderFile struct {
	Type          string   `json:"type"`
	Path          string   `json:"path"`
	Visibility    string   `json:"visibility"`
	LastModified  int64    `json:"last_modified"`
	MimeType      string   `json:"mime_type"`
	ExtraMetadata []string `json:"extra_metadata"`
	Basename      string   `json:"basename"`
	Storage       string   `json:"storage"`
	FileSize      int64    `json:"file_size"`
}

type VuefinderFileResponse struct {
	Adapter  string          `json:"adapter"`
	Storages []string        `json:"storages"`
	Dirname  string          `json:"dirname"`
	Files    []VuefinderFile `json:"files"`
}

func createVuefinderFiles(adapter, pth string, files []fs.DirEntry) []VuefinderFile {
	vuefinderFiles := []VuefinderFile{}

	for _, ff := range files {
		if fi, err := ff.Info(); err == nil {
			vueFinderFile := VuefinderFile{
				Path:         adapter + "://" + path.Join(pth, fi.Name()),
				Visibility:   "public",
				LastModified: fi.ModTime().UnixMilli(),
				MimeType:     mime.TypeByExtension(path.Ext(fi.Name())),
				Basename:     fi.Name(),
				Storage:      adapter,
				FileSize:     fi.Size(),
			}

			if fi.IsDir() {
				vueFinderFile.Type = "dir"
			} else {
				vueFinderFile.Type = "file"
			}

			vuefinderFiles = append(vuefinderFiles, vueFinderFile)
		}
	}

	return vuefinderFiles
}

func getFiles(c *fiber.Ctx) responseMessage {
	var response responseMessage

	// if there is no adapter specified, use PUBLIC
	adapter := c.Query("adapter", "PUBLIC")

	if adapter == "null" || adapter == "undefined" {
		adapter = "PUBLIC"
	}

	adapterPath := extractPath(c.Query("path"), adapter)

	if pth, err := Config.sanitizeUploadDir(adapterPath); err != nil {
		response.Status = fiber.StatusBadRequest
	} else {
		if res, err := fs.ReadDir(Config.UploadDirSys, pth); err != nil {
			response.Status = fiber.StatusBadRequest
		} else {
			response.Data = VuefinderFileResponse{
				Adapter:  adapter,
				Storages: []string{adapter},
				Dirname:  adapterPath,
				Files:    createVuefinderFiles(adapter, pth, res),
			}
		}
	}

	return response
}

func getPreview(c *fiber.Ctx) responseMessage {
	var response responseMessage
	var err error

	// check wether all query arguments are available
	if pth := c.Query("path"); pth == "" {
		response.Status = fiber.StatusBadRequest
	} else if adapter := c.Query("adapter"); adapter == "" {
		response.Status = fiber.StatusBadRequest
	} else if pth, err = Config.getUploadDir(extractPath(pth, adapter)); err != nil {
		response.Status = fiber.StatusBadRequest
	} else if buffer, err := os.ReadFile(pth); err != nil {
		response.Status = fiber.StatusBadRequest
	} else {
		response.Buffer = buffer
	}

	return response
}

func getSubfolders(c *fiber.Ctx) responseMessage {
	var response responseMessage

	// check wether all query arguments are available
	if adapterPath := c.Query("path"); adapterPath == "" {
		response.Status = fiber.StatusBadRequest
	} else if adapter := c.Query("adapter"); adapter == "" {
		response.Status = fiber.StatusBadRequest
	} else if pth, err := Config.sanitizeUploadDir(extractPath(adapterPath, adapter)); err != nil {
		response.Status = fiber.StatusBadRequest
	} else if res, err := fs.ReadDir(Config.UploadDirSys, pth); err != nil {
		response.Status = fiber.StatusBadRequest
	} else {
		// remove non-directories from the files
		removeCount := 0

		for ii, ff := range res {
			if !ff.IsDir() {
				removeCount++

				res[ii] = res[len(res)-removeCount]
			}

		}
		response.Data = struct {
			Folders []VuefinderFile `json:"folders"`
		}{
			Folders: createVuefinderFiles(adapter, pth, res[:len(res)-removeCount]),
		}
	}

	return response
}

func getDownload(c *fiber.Ctx) responseMessage {
	return getPreview(c)
}

func postNewFolder(c *fiber.Ctx) responseMessage {
	var response responseMessage

	// check wether all query arguments are available
	if adapter := c.Query("adapter"); adapter == "" {
		response.Status = fiber.StatusBadRequest
	} else {
		body := new(struct {
			Name string `json:"name"`
		})

		pth := extractPath(c.Query("path"), adapter)

		if err := c.BodyParser(&body); err != nil {
			logger.Warn(`"body" can't be parsed as "{ name string }"`)
			response.Status = fiber.StatusBadRequest
		} else if pth, err := Config.getUploadDir(path.Join(pth, body.Name)); err != nil {
			response.Status = fiber.StatusBadRequest
		} else if err := os.Mkdir(pth, fs.ModeDir|0777); err != nil {
			response.Status = fiber.StatusBadRequest
		} else {
			response = getFiles(c)
		}
	}

	return response
}

func postRename(c *fiber.Ctx) responseMessage {
	var response responseMessage

	// check wether all query arguments are available
	if adapter := c.Query("adapter"); adapter == "" {
		response.Status = fiber.StatusBadRequest
	} else {
		body := new(struct {
			Item string `json:"item"`
			Name string `json:"name"`
		})

		if err := c.BodyParser(&body); err != nil {
			logger.Warn(`"body" can't be parsed as "{ item string; name string }"`)
			response.Status = fiber.StatusBadRequest
		} else if pthOld, err := Config.getUploadDir(extractPath(body.Item, adapter)); err != nil {
			response.Status = fiber.StatusBadRequest
		} else {
			pthNew := path.Join(path.Dir(pthOld), body.Name)

			if err := os.Rename(pthOld, pthNew); err != nil {
				response.Status = fiber.StatusBadRequest
			} else {
				response = getFiles(c)
			}
		}
	}

	return response
}

func postMove(c *fiber.Ctx) responseMessage {
	var response responseMessage

	// check wether all query arguments are available
	if adapter := c.Query("adapter"); adapter == "" {
		response.Status = fiber.StatusBadRequest
	} else {
		body := new(struct {
			Item  string `json:"item"`
			Items []struct {
				Path string `json:"path"`
			} `json:"items"`
		})

		if err := c.BodyParser(&body); err != nil {
			logger.Warn(`"body" can't be parsed as "{ item string; items []{ path string } }"`)
			response.Status = fiber.StatusBadRequest
		} else if destDir, err := Config.sanitizeUploadDir(extractPath(body.Item, adapter)); err != nil {
			response.Status = fiber.StatusBadRequest
		} else {
			for _, ff := range body.Items {
				// validate the new path
				if orig, err := Config.getUploadDir(extractPath(ff.Path, adapter)); err == nil {
					if dest, err := Config.getUploadDir(path.Join(destDir, path.Base(orig))); err == nil {
						if err := os.Rename(orig, dest); err != nil {
							response.Status = fiber.StatusBadRequest

							break
						}
					} else {
						response.Status = fiber.StatusBadRequest

						break
					}
				} else {
					response.Status = fiber.StatusBadRequest

					break
				}
			}

			response = getFiles(c)
		}
	}

	return response
}

func postDelete(c *fiber.Ctx) responseMessage {
	var response responseMessage

	// check wether all query arguments are available
	if adapter := c.Query("adapter"); adapter == "" {
		response.Status = fiber.StatusBadRequest
	} else {
		body := new(struct {
			Items []struct {
				Path string `json:"path"`
			} `json:"items"`
		})

		if err := c.BodyParser(&body); err != nil {
			logger.Warn(`"body" can't be parsed as "{ item string; items []{ path string } }"`)
			response.Status = fiber.StatusBadRequest
		} else {
			for _, ff := range body.Items {
				// validate the new path
				if pth, err := Config.getUploadDir(extractPath(ff.Path, adapter)); err == nil {
					if err := os.Remove(pth); err == nil {
						response.Status = fiber.StatusBadRequest

						break
					}
				} else {
					response.Status = fiber.StatusBadRequest

					break
				}
			}

			response = getFiles(c)
		}
	}

	return response
}

func init() {
	endpoints := map[string]map[string]func(*fiber.Ctx) responseMessage{
		"GET": {
			"index":      getFiles,
			"preview":    getPreview,
			"subfolders": getSubfolders,
			"download":   getDownload,
		},
		"POST": {
			"newfolder": postNewFolder,
			"rename":    postRename,
			"move":      postMove,
			"delete":    postDelete,
		},
	}

	handleMethods := map[string]func(path string, handlers ...func(*fiber.Ctx) error) fiber.Router{
		"GET":  app.Get,
		"POST": app.Post,
	}

	for method, handler := range handleMethods {
		handler("/api/storage/browse", func(c *fiber.Ctx) error {
			logger.Sugar().Debugf("HTTP %s request: %q", c.Method(), c.OriginalURL())

			var response responseMessage

			// check wether the session-cookie is valid and the user is an admin
			if admin, err := checkAdmin(c); err != nil {
				response.Status = fiber.StatusInternalServerError

				logger.Sugar().Errorf("can't check for admin: %s", err)
			} else if !admin {
				response.Status = fiber.StatusUnauthorized

				// check for a valid query
			} else if c.Query("q") == "" {
				response.Status = fiber.StatusBadRequest

				logger.Sugar().Info(`query is missing "q"`)
			} else if requestHandler, ok := endpoints[method][c.Query("q")]; !ok {
				response.Status = fiber.StatusBadRequest

				logger.Sugar().Infof(`invalid value for "q" in query: %s`, c.Query("q"))
			} else {
				response = requestHandler(c)
			}

			return response.send(c)
		})
	}
}
